src/sql/rewrite/ob_transform_simplify_expr.cpp

```diff
@@ -2517,4 +2526,740 @@ int ObTransformSimplifyExpr::try_remove_ora_decode(ObRawExpr *&expr,
+ * @discription: according to the results of when exprs
+ *  1. cond1 ... cond(k-1) is FALSE/NULL
+ *  2. condk is TRUE/UNCALCULABLE
+ *    2.1 if condk is TRUE, then rewrite case_expr to exprk
+ *    2.2 if condk is UNCALCULABLE, then remove cond1...cond(k-1)
+ *    2.3 if k > when_cnt, then rewrite case_expr to default_expr
+ */
+int ObTransformSimplifyExpr::convert_case_when_predicate_by_when_exprs(
+                                                      ObRawExpr *&expr, 
+                                                      bool &trans_happened) {
+  int ret = OB_SUCCESS;
+
+  trans_happened = false;
+  ObCaseOpRawExpr *case_expr = NULL;
+  bool has_null = false; // check cond1 to cond(k-1) has null expr or not
+  bool is_true = false;  // check condk is true expr or not
+  bool is_uncalculable = false; // check condk is uncalculable or not
+  int64_t first_true_non_calc_idx = -1; // record the index of first true or non_calc expr
+  ObSEArray<ObRawExpr*, 2> false_null_exprs; // used to add constraints
+  bool is_valid = false;
+
+  if (OB_ISNULL(expr) || OB_ISNULL(ctx_)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("param is NULL", K(ret));
+  } else if (T_OP_CASE != expr->get_expr_type()) {
+  } else if (OB_FALSE_IT(case_expr = static_cast<ObCaseOpRawExpr *>(expr))) {
+  } else if (OB_UNLIKELY(case_expr->get_when_expr_size() != case_expr->get_then_expr_size())) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("incorrect case when params", K(ret));
+  } else if (OB_UNLIKELY(case_expr->get_when_expr_size() < 1)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("case when size is less than 1", K(ret));
+  } else if (OB_FAIL(check_convert_case_when_by_when_exprs_validity(
+                                                case_expr,
+                                                has_null, 
+                                                is_true,
+                                                is_uncalculable,
+                                                first_true_non_calc_idx,
+                                                false_null_exprs,
+                                                is_valid
+                                                ))) {
+    LOG_WARN("failed to check inner convert case when validity");
+  } else if (!is_valid) {
+    /* do nothing */
+  } else { // rewrite case_expr and add constraint
+    int64_t false_null_cnt = false_null_exprs.count();
+    int64_t when_cnt = case_expr->get_when_expr_size();
+    // collect rewrite expr
+    ObRawExpr *rewrite_expr = NULL;
+    ObSEArray<ObRawExpr *, 2> when_exprs;
+    ObSEArray<ObRawExpr *, 2> then_exprs;
+    bool has_trans = true;
+    if (OB_LIKELY(-1 == first_true_non_calc_idx)) {
+      ObRawExpr *default_expr = NULL;
+      if (OB_ISNULL(default_expr = case_expr->get_default_param_expr())) {
+        ret = OB_ERR_UNEXPECTED;
+        LOG_WARN("default expr is NULL", K(ret));
+      } else if (OB_UNLIKELY(false_null_cnt != when_cnt)) {
```

Comment:
convert_case_when_predicate_by_when_exprs is too complex.
try convert the first when directly, and do this convert for the same expr repeatedly if transform happened.